poh/src/poh_service.rs

```diff
@@ -357,54 +462,100 @@ impl PohService {
-                hashes_per_batch,
-                &poh,
-                target_ns_per_tick,
-            );
-            if should_tick {
-                // Lock PohRecorder only for the final hash. record_or_hash will lock PohRecorder for record calls but not for hashing.
-                {
-                    let mut lock_time = Measure::start("lock");
-                    let mut poh_recorder_l = poh_recorder.write().unwrap();
-                    lock_time.stop();
-                    timing.total_lock_time_ns += lock_time.as_ns();
-                    let mut tick_time = Measure::start("tick");
-                    poh_recorder_l.tick();
-                    tick_time.stop();
-                    timing.total_tick_time_ns += tick_time.as_ns();
+            // If we should exit, close the channel so no more records are accepted,
+            // but we still want to process any pending records.
+            // We should **not** however process any service messages once we have detected
+            // the exit signal.
+            should_exit |= poh_exit.load(Ordering::Relaxed); // once set, stay set.
+            if should_exit {
+                record_receiver.shutdown();
+            }
+
+            let service_message =
+                Self::check_for_service_message(&poh_service_receiver, &mut record_receiver);
+            loop {
+                let should_tick = Self::record_or_hash(
+                    &mut next_record,
+                    &poh_recorder,
+                    &mut timing,
+                    &mut record_receiver,
+                    hashes_per_batch,
+                    &poh,
+                    target_ns_per_tick,
+                    ticks_per_slot,
+                );
+                if should_tick {
+                    // Lock PohRecorder only for the final hash. record_or_hash will lock PohRecorder for record calls but not for hashing.
+                    {
+                        let mut lock_time = Measure::start("lock");
+                        let mut poh_recorder_l = poh_recorder.write().unwrap();
+                        lock_time.stop();
+                        timing.total_lock_time_ns += lock_time.as_ns();
+                        let mut tick_time = Measure::start("tick");
+                        poh_recorder_l.tick();
+                        tick_time.stop();
+                        timing.total_tick_time_ns += tick_time.as_ns();
+                    }
+                    timing.num_ticks += 1;
+
+                    timing.report(ticks_per_slot);
                 }
-                timing.num_ticks += 1;
 
-                timing.report(ticks_per_slot);
-                if poh_exit.load(Ordering::Relaxed) {
+                // If there is a service message (channel is shut down), and there
+                // still records to process, we must loop to call `record_or_hash`
+                // again.
+                if service_message.is_some() && !record_receiver.is_empty() {
```

Comment:
what are the implications of the receiver not being empty while we change the poh bank? 
In AG world:
- No other thread messes with poh_recorder other than BlockCreation (so in theory PohController is not needed)
- BlockCreation will spin record_or_hash, when certain conditions are reached it will manually alpentick and insert a new poh bank
Does there need to be synchrony here? what if there is a new record inbetween the last record_or_hash and inserting a new poh bank?

Reply:
Gotcha! I was planning on trying to remove the non-production versions of PoH-service. They test behavior we don't need; tests that rely on PoH service mostly don't actually need it.

If it's useful for AG, I can leave them for you though